src/h5web/visualizations/heatmap/HeatmapVis.tsx

```diff
@@ -30,24 +36,66 @@ function HeatmapVis(props: Props): JSX.Element {
     keepAspectRatio = true,
     showGrid = false,
     showLoader = true,
+    abscissas,
+    ordinates,
   } = props;
 
   const { rows, cols } = getDims(dataArray);
   const aspectRatio = keepAspectRatio ? cols / rows : undefined; // width / height <=> cols / rows
 
+  const abscissaIndices = range(cols + 1);
+  const abscissaValues = abscissas || abscissaIndices;
+  const abscissaToIndex = getValueToIndexScale(abscissaValues);
+  const abscissaDomain = useMemo(() => {
+    return getDataDomain(abscissaValues) ?? DEFAULT_DOMAIN;
+  }, [abscissaValues]);
+
+  const ordinateIndices = range(rows + 1);
+  const ordinateValues = ordinates || ordinateIndices;
+  const ordinateToIndex = getValueToIndexScale(ordinateValues);
+  const ordinateDomain = useMemo(() => {
+    return getDataDomain(ordinateValues) ?? DEFAULT_DOMAIN;
+  }, [ordinateValues]);
+
+  if (abscissaValues.length !== cols + 1) {
+    return (
+      <p>{`Abscissas size (${
+        abscissaValues.length
+      }) does not match data length along X (${cols + 1})`}</p>
+    );
+  }
+
+  if (ordinateValues.length !== rows + 1) {
+    return (
+      <p>{`Ordinate size (${
+        ordinateValues.length
+      }) does not match data length along Y (${rows + 1})`}</p>
+    );
+  }
+
   return (
     <div className={styles.root}>
       <VisCanvas
-        abscissaConfig={{ indexDomain: [0, cols], showGrid }}
-        ordinateConfig={{ indexDomain: [0, rows], showGrid }}
+        abscissaConfig={{
+          domain: abscissaDomain,
+          showGrid,
+          isIndexAxis: !abscissas,
+        }}
+        ordinateConfig={{
+          domain: ordinateDomain,
+          showGrid,
+          isIndexAxis: !ordinates,
+        }}
         aspectRatio={aspectRatio}
       >
         <TooltipMesh
-          formatIndex={([x, y]) => `x=${Math.floor(x)}, y=${Math.floor(y)}`}
+          formatIndex={([x, y]) => {
+            return `x=${abscissaToIndex(x)}, y=${ordinateToIndex(y)}`;
+          }}
           formatValue={([x, y]) => {
-            return x < cols && y < rows
-              ? format('.3')(dataArray.get(Math.floor(y), Math.floor(x)))
-              : undefined;
+            return format('.3')(
+              dataArray.get(ordinateToIndex(y), abscissaToIndex(x))
+            );
```

Comment:
Now that domains of axis can be arbitrary values, the tooltip coordinates are no longer indices. To retrieve the tooltip value, I need a way to get the index from the value and this is done with a threshold scale that maps values to indices.

See the comment below for more details on the scale construction.